api/tools/src/main/java/org/pmiops/workbench/tools/FixDesynchronizedBillingProjectOwners.java

```diff
@@ -0,0 +1,202 @@
+
+  private static void dryLog(boolean dryRun, String msg) {
+    String prefix = "";
+    if (dryRun) {
+      prefix = "[DRY RUN] Would have... ";
+    }
+    log.info(prefix + msg);
+  }
+
+  private static void clean(
+      WorkspacesApi workspacesApi,
+      BillingApi billingApi,
+      Set<String> billingProjectIds,
+      String researcherDomain,
+      boolean dryRun)
+      throws ApiException {
+    int ownersRemoved = 0;
+    int ownersAdded = 0;
+    for (FirecloudWorkspaceResponse resp :
+        workspacesApi.listWorkspaces(FIRECLOUD_LIST_WORKSPACES_REQUIRED_FIELDS)) {
+      FirecloudWorkspace w = resp.getWorkspace();
+      if (!billingProjectIds.isEmpty() && !billingProjectIds.contains(w.getNamespace())) {
+        continue;
+      }
+
+      String id = w.getNamespace() + "/" + w.getName();
+      if (!"PROJECT_OWNER".equals(resp.getAccessLevel())) {
+        log.warning(
+            String.format(
+                "service account has '%s' access to workspace '%s'; skipping",
+                resp.getAccessLevel(), id));
+        continue;
+      }
+
+      Map<String, String> billingProjectRoles =
+          billingApi.listBillingProjectMembers(w.getNamespace()).stream()
+              .filter(m -> m.getEmail().endsWith("@" + researcherDomain))
+              .collect(
+                  Collectors.toMap(
+                      FirecloudBillingProjectMember::getEmail,
+                      FirecloudBillingProjectMember::getRole));
+      Set<String> billingProjectOwners =
+          billingProjectRoles.entrySet().stream()
+              .filter(e -> "Owner".equals(e.getValue()))
+              .map(Entry::getKey)
+              .collect(Collectors.toSet());
+
+      Map<String, String> workspaceRoles =
+          FirecloudTransforms.extractAclResponse(
+                  workspacesApi.getWorkspaceAcl(w.getNamespace(), w.getName()))
+              .entrySet().stream()
+              .filter(e -> e.getKey().endsWith("@" + researcherDomain))
+              .collect(Collectors.toMap(Entry::getKey, e -> e.getValue().getAccessLevel()));
+      Set<String> workspaceOwners =
+          workspaceRoles.entrySet().stream()
+              .filter(e -> "OWNER".equals(e.getValue()))
+              .map(e -> e.getKey())
+              .collect(Collectors.toSet());
+
+      // ShareWorkspace first updates the Workspace ACL, then the billing project role. For this
+      // reason, the workspace ACL is the source of truth, as it will have been set properly in the
+      // event of a partial sharing failure.
+      for (String user : Sets.difference(billingProjectOwners, workspaceOwners)) {
+        dryLog(
+            dryRun,
+            String.format(
+                "'%s': '%s', inconsistency: workspace role '%s', project role '%s'",
+                w.getNamespace(), user, workspaceRoles.get(user), billingProjectRoles.get(user)));
+        if (billingProjectOwners.contains(user)) {
```

Comment:
wouldn't this always be true if `user` is in the set of `billingProjectOwners` - `workspaceOwners`? `user` is always an element within `billingProjectOwners` which is not in `workspaceOwners`

Reply:
Great catch - thanks. There were two bugs here:

1. I accidentally imported from the Elasticsearch client library... that was not intentional. Should have been Guava.
2. I'd thought `Sets.difference` was the symmetric difference. Fixed to actually call the `symmetricDifference`, which was my intent.